### PR TITLE
feat: add minimum available capacity slack concept

### PR DESCRIPTION
**Pull request by miguelsorianod**
## Description

As part of dynamic scaling, the concept of a minimum
available capacity slack in streaming units has been introduced.
The available capacity slack of a supported instance type in a region
is the desired minimum number of streaming units that should
be available(free) at any moment for that instance type in the region.

It has been introduced in the provider configuration file of the
fleet manager. There is one capacity slack value per defined instance
type in the region.
The attribute is optional. If it is not set then it is interpreted
as a capacity slack of 0 is desired.
The capacity slack is intended to be used only when dynamic scaling
is enabled.

The capacity slack is defined as the `min_available_capacity_slack_streaming_units` yaml attribute  

This will require app-interface changes. The desired value for `min_available_capacity_slack_streaming_units` is to be determined. The app-interface changes should be done when enabling dynamic scaling functionality, along with all the other dynamic scaling related attributes

## Verification Steps

* Add `min_available_capacity_slack_streaming_units` attribute in the providers configuration for standard and/or developer instance with a value of `-1`. An error should be returned
* Add the `limit` attribute in the providers config with a valid value for an instance type and also the `min_available_capacity_slack_streaming_units` with a bigger value than the value set to `limit`. An error should be returned. Repeat with all the instance types in the provider config
* Add  the `limit` attribute in the provider config with a valid value and a valid value for `min_available_capacity_slack_streaming_units` being smaller than the `limit` value. fleet manager should start correctly (as long as the cluster level metrics add up to the defined region level limits)
* Verify that if dynamic scaling flag is enabled the data plane config with the defined clusters information is not taken into account
* Verify that if the `min_available_capacity_slack_streaming_units` value is not defined an error is not returned (0 is assumed so in case a region limit is configured it will never be bigger than the limit value)

## Checklist (Definition of Done)

- [ ] All acceptance criteria specified in JIRA have been completed
- [ ] Unit and integration tests added that prove the fix is effective or the feature works (tested against emulated and non-emulated OCM environment)
- [ ] Documentation added for the feature
- [ ] CI and all relevant tests are passing
- [ ] Code Review completed
- [ ] Verified independently by reviewer
- [ ] All PR comments are resolved either by addressing them or creating follow up tasks
- [ ] Required metrics/dashboards/alerts have been added (or PR created).
- [ ] Required Standard Operating Procedure (SOP) is added.
- [ ] JIRA has been created for changes required on the client side

### CODE REVIEW

**Review by JameelB — COMMENTED** [review]
lgtm, can you add documnentation and example for this in the [provider-configuration.yaml](https://github.com/bf2fc6cc711aee1a0c2a/kas-fleet-manager/blob/main/config/provider-configuration.yaml)? Adding it to the comments at the top of should be fine.

**Comment by miguelsorianod** [comment]
> lgtm, can you add documnentation and example for this in the [provider-configuration.yaml](https://github.com/bf2fc6cc711aee1a0c2a/kas-fleet-manager/blob/main/config/provider-configuration.yaml)? Adding it to the comments at the top of should be fine.

:+1: I will add documentation in the config file itself

**Comment by miguelsorianod** [comment]
@JameelB I updated the config file with the new attribute information and also a little bit of rewording of the old doc

**Comment by machi1990** [comment]
The change looks good to me.
There is a related test failure
```
=== FAIL: cmd/kas-fleet-manager TestInjections (0.00s)
    main_test.go:22: 
        Expected
            <*errors.errorString | 0xc0005a6a00>: {
                s: "limit for developer instance type (3) does not match cluster capacity configuration in region 'us-east-1': min(0), max(0)",
            }
        to be nil
```
can you have a look?

**Comment by miguelsorianod** [comment]
> The change looks good to me. There is a related test failure
> 
> ```
> === FAIL: cmd/kas-fleet-manager TestInjections (0.00s)
>     main_test.go:22: 
>         Expected
>             <*errors.errorString | 0xc0005a6a00>: {
>                 s: "limit for developer instance type (3) does not match cluster capacity configuration in region 'us-east-1': min(0), max(0)",
>             }
>         to be nil
> ```
> 
> can you have a look?

Fixed. The tests should pass now

**Review by machi1990 — APPROVED** [review]

File: config/provider-configuration.yaml [hunk 1]
```diff
@@ -1,26 +1,40 @@
...
+#   ...
+#
+#   With the above configuration, up to 5 'standard' Kafka instances can be created in the us-east-1 region. Since there is no region limit set for 'developer',
+#   KAS Fleet Manager will allow creation of 'developer' Kafka instances in the us-east-1 region as long as a data plane cluster is available in this region
+#   (i.e. is schedulable, has remaining capacity and supports this Kafka instance type)
+#
+#   Note: If manual scaling is enabled, please ensure that the limits you configure here matches/must not exceed the kafka_instance_limits of your data
+#         plane clusters in the dataplane-cluster-configuration.yaml file
+#
+# min_available_capacity_slack_streaming_units:
```
Comment (JameelB) [inline config/provider-configuration.yaml:37]: Thanks for updating this! Is this meant to have additional info or can it be removed?

Comment (miguelsorianod) [reply config/provider-configuration.yaml:37]: that should be removed. I will update the PR

Comment (miguelsorianod) [reply config/provider-configuration.yaml]: updated

**Pull request merged**